acdc-ws/app/services/Metric.scala

```diff
@@ -0,0 +1,99 @@
+import javax.inject.{Inject, Singleton}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+import com.typesafe.config.ConfigFactory
+import play.api.Configuration
+import play.api.mvc.RequestHeader
+import io.prometheus.client._
+import io.prometheus.client.exporter.common.TextFormat
+import io.prometheus.client.hotspot.DefaultExports
+
+trait Metric {
+  private lazy val config = new Configuration(ConfigFactory.load())
+
+  private val disableMetrics =
+    config.getOptional[String]("acdc.metrics.endpoint").getOrElse("").isEmpty
+
+  private val bypassPaths = config.getOptional[Seq[String]]("acdc.metrics.bypass.paths") match {
+    case Some(paths) => paths.toSet
+    case _ => Set[String]()
+  }
+
+  private val STATIC_PATH_MARKERS = Seq("instance", "dataset", "lineage", "__metrics", "__status")
+
+  def parseRequest(request: RequestHeader): Option[(String, String)] = {
+    if (checkPathIsDisabled(request.path) || checkMetricIsDisabled)
+      None
+    else {
+      val (staticPath, args) = request.path
+        .split("/")
+        .filter(_.nonEmpty)
+        .foldLeft((List[String](), List[String]())) {
+          case ((Nil, l2), token) => (List(token), l2)
+          case ((h :: t, l2), token) =>
+            if (STATIC_PATH_MARKERS.contains(h)) (h :: t, token :: l2)
+            else (token :: h :: t, l2)
+        }
+      Some((staticPath.reverse.mkString("-"), args.reverse.mkString("-")))
+    }
+  }
+
+  def checkMetricIsDisabled: Boolean = { disableMetrics }
+
+  def checkPathIsDisabled(path: String): Boolean = { bypassPaths.contains(path) }
+
+  def incrementStatusCount(status: String): Unit
+
+  def startApiTimer(labels: String*): () => Unit
+
+  def collect: Future[String]
+
+  def clear(): Unit
+
+}
+
+@Singleton
+class PrometheusMetric @Inject() (implicit ec: ExecutionContext) extends Metric {
+
+  DefaultExports.initialize()
+
+  private val httpStatusCount: Counter = Counter.build
+    .name("http_requests_total")
+    .help("Total HTTP Requests Count")
+    .labelNames("status")
+    .register
+
+  private val apiLatency: Summary = Summary
+    .build()
+    .name("apiLatencySummary")
+    .labelNames("path", "arguments", "method")
+    .help("Profile API response time summary")
+    .quantile(0.5d, 0.001d)
+    .quantile(0.95d, 0.001d)
+    .quantile(0.99d, 0.001d)
+    .register
+
+  override def incrementStatusCount(status: String): Unit = {
+    httpStatusCount.labels(status).inc()
+  }
+
+  override def startApiTimer(labels: String*): () => Unit = {
+    val timer = apiLatency.labels(labels: _*).startTimer()
+    val callback = () => timer.close
+    callback
+  }
+
+  // Get metrics from the local prometheus collector default registry
+  override def collect: Future[String] = Future {
+    val writer = new StringWriter()
+    TextFormat.write004(writer, CollectorRegistry.defaultRegistry.metricFamilySamples())
+    writer.toString
+  }
+
+  override def clear(): Unit = { apiLatency.clear() }
```

Comment:
@thomaschangsf I think we do not have the clear the apiLatency.  The summary default behavior  has a sliding window of 10 mins with step size 2 mins.  So the percentiles would be relevant with recency already.  

The clear would reset the count and sum to zero, I think.  And we would lose the global average since the service started upon each scrape.  Given that the recent times statistics is available in the percentiles, maybe not reset the sum / count so to have info on the longer time average.

Reply:
@schowsf 
Thank you for sharing the nuances of the clear method and posing this question.

Clearing the count and sum after each metric scrape would indeed lose our global average.  For the READ dashboard, what is the value of having a global count/sum to the user?  Perhaps it is give a historical comparison?

I can go either way as long as we are consistent across all our services.